### PR TITLE
Add implementation for TransformerFactory XXE recipe

### PR DESCRIPTION
**Pull request by aaronist**
- Add functionality for fixing XXE vulnerabilities with `TransformerFactory`

### CODE REVIEW

File: src/test/java/org/openrewrite/java/security/TransformerFactoryXXEVulnerabilityTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ */
+package org.openrewrite.java.security;
+
+import org.junit.jupiter.api.Test;
+import org.openrewrite.test.RecipeSpec;
+import org.openrewrite.test.RewriteTest;
+
+import static org.openrewrite.java.Assertions.java;
+
+public class TransformerFactoryXXEVulnerabilityTest implements RewriteTest {
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/security/TransformerFactoryXXEVulnerabilityTest.java:24]:
```suggestion
class TransformerFactoryXXEVulnerabilityTest implements RewriteTest {
```

**Comment by timtebeek** [comment]
Hi @aaronist ; Could you give me bit more background on the state and intention of this pull request? That helps me review before a merge.

I'm seeing two visitors added here, but no accompanying recipe. To run these either through OpenRewrite or Moderne I'd expect a recipe to go along with these visitors, unless that recipe should live in a different repository. What are your plans with these visitors?

**Comment by aaronist** [comment]
Hi @timtebeek, this draft pull request is still a WIP; I created it so some of my peers could examine my work. Sorry if that caused any confusion on your end.

The aim is to continue expanding remediation of XXE vulnerabilities by tackling the `TransformerFactory` parser. The recipe in question is likely going to be some abstraction of the `XmlParserXXEVulnerability` class (which is yet to be implemented). I'll continue to work on this PR over the next weeks to make sure it's fully fleshed out.

**Comment by JLLeitschuh** [comment]
> Alternatively we can also merge and test it out through the platform from the snapshot, but that might require some coordination for the recipe deployments.

This was what my plans were

**Comment by JLLeitschuh** [comment]
> I'm seeing two visitors added here, but no accompanying recipe. To run these either through OpenRewrite or Moderne I'd expect a recipe to go along with these visitors, unless that recipe should live in a different repository. What are your plans with these visitors?

They will be run as a part of existing recipes

**Comment by timtebeek** [comment]
Sure! Just let me know when this is ready for review; after the first merged PRs I won't have to approve the CI pipelines anymore, but until then feel free to ping me if that's ever needed.

File: src/test/java/org/openrewrite/java/security/xml/TransformerFactoryXXEVulnerabilityTest.java [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+package org.openrewrite.java.security.xml;
+
+import org.junit.jupiter.api.Test;
+import org.openrewrite.java.security.XmlParserXXEVulnerability;
+import org.openrewrite.test.RecipeSpec;
+import org.openrewrite.test.RewriteTest;
+
+import static org.openrewrite.java.Assertions.java;
+
+class TransformerFactoryXXEVulnerabilityTest implements RewriteTest {
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/security/xml/TransformerFactoryXXEVulnerabilityTest.java:25]: Should we rename this class _and file_ to match the recipe it tests?
```suggestion
class XmlParserXXEVulnerabilityTest implements RewriteTest {
```

File: src/main/java/org/openrewrite/java/security/XmlParserXXEVulnerability.java [hunk 3]
```diff
@@ -47,7 +50,15 @@ public TreeVisitor<?, ExecutionContext> getScanner(ExternalDTDAccumulator acc) {
     //q: What is Java doc format?
     @Override
     public TreeVisitor<?, ExecutionContext> getVisitor(ExternalDTDAccumulator acc) {
-        return XmlInputFactoryFixVisitor.create(acc);
+        return new TreeVisitor<Tree, ExecutionContext>() {
+            @Override
+            public Tree visit(@Nullable Tree tree, ExecutionContext executionContext) {
+                if (!(tree instanceof Xml)) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/security/XmlParserXXEVulnerability.java:56]: Why (only) the exclusion of XML here, versus for instance only passing through Java (+ Groovy/Kotlin ?) source files?

File: src/main/java/org/openrewrite/java/security/xml/TransformerFactoryInsertAttributeStatementVisitor.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    public J.Block visitBlock(J.Block block, P ctx) {
+        J.Block b = super.visitBlock(block, ctx);
+        Statement beforeStatement = null;
+        if (b.isScope(scope)) {
+            for (int i = b.getStatements().size() - 2; i > -1; i--) {
+                Statement st = b.getStatements().get(i);
+                Statement stBefore = b.getStatements().get(i + 1);
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/security/xml/TransformerFactoryInsertAttributeStatementVisitor.java:58]: Getting these statements by offset seems somewhat strange to me; how certain are we that they always occur in a certain order?

**Review by timtebeek — COMMENTED** [review]
Some questions as I went through the PR; I imagine @JLLeitschuh will want to have a look as well.

File: src/main/java/org/openrewrite/java/security/xml/TransformerFactoryInsertAttributeStatementVisitor.java (hunk 4 above)
Comment (JLLeitschuh) [reply src/main/java/org/openrewrite/java/security/xml/TransformerFactoryInsertAttributeStatementVisitor.java:58]: This is copied from `XmlFactoryInsertPropertyStatementVisitor`, which was already written.

File: src/test/java/org/openrewrite/java/security/xml/TransformerFactoryXXEVulnerabilityTest.java (hunk 2 above)
Comment (JLLeitschuh) [reply src/test/java/org/openrewrite/java/security/xml/TransformerFactoryXXEVulnerabilityTest.java:26]: There is already a `XmlParserXXEVulnerabilityTest`

Comment (JLLeitschuh) [reply src/test/java/org/openrewrite/java/security/xml/TransformerFactoryXXEVulnerabilityTest.java:26]: This test is actually testing an individual visitor, which the recipe happens to invoke. Instead, we chose to rename the `XmlInputFactoryXXEVulnerabilityTest`

**Review by JLLeitschuh — APPROVED** [review]

**Pull request merged**